### DISCUSSION TITLE
No way to trigger unsaved changes dialogs with `wire:navigate`

### DISCUSSION THREAD

**Discussion by binaryfire** [livewire/livewire#7734] [1. Help] [answered]
Currently there's no way to trigger an 'unsaved changes' dialog / modal when navigating away from a page with `wire:navigate`.  This is because the `beforeunload` event isn't triggered. Other JS frameworks handle this by providing a custom guard [eg. vue-router](https://router.vuejs.org/guide/advanced/navigation-guards.html):

<img width="566" alt="image" src="https://github.com/livewire/livewire/assets/5361908/e38e96cf-bd98-4b29-8188-acaa86549814">

`livewire:navigating` is the perfect event for checking for unsaved changes and triggering a custom dialog, but there's the no way to break the `wire:navigate` request and stop it from happening. It would be great if you could add a mechanism for doing this. Perhaps an event we can fire to cancel navigation?

Filament has implemented custom 'unsaved changes' functionality but there's no way to make it work in 'wire:navigate' mode because of this issue. See: https://github.com/filamentphp/filament/pull/10516#issuecomment-1890708436

**Comment by binaryfire** [comment]
Does anyone have a suggestion for a workaround? We’re migrating a healthcare app from a Vue SPA to Livewire because of `wire:navigate` and now can’t warn users about unsaved changes if they click another link before saving.

It’s a major issue in this industry (and I imagine several others). Autosave isn’t a viable alternative - even 30 sec of lost input isn’t acceptable for a health professional. It needs to be a confirmation modal or dialog, ie. something that forces user interaction.

**Comment by moduleitor** [comment]
Hello! I know... late, but...

Maybe combining wire:dirty and Alpine.navigate, instead of using wire:navigate directly?

I haven't tried it, but it sounds like a possible partial solution...

Any suggestions from someone with more experience with Livewire 3 and wire:navigate?

**Reply by binaryfire** [reply]
Using ‘Alpine.navigate’ instead is an interesting idea. Trying to intercept ‘wire:navigate’ was a losing battle. Would be great to get some feedback from the Livewire / Alpine gurus before I go down another rabbit hole though 😅

**Comment by codewithluck** [comment]
I have the same issue with the topic's owner and I am looking around for a solution. In jQuery I can use off() event to stop all event listeners, but in AlpineJS it is impossible, I have many navigation items and other links on the app that make it not easy to remove all wire:navigate. Can anyone help, please?

**Comment by binaryfire** [comment]
@codewithluck We had to drop our Vue -> Livewire migration plans for a few mission-critical apps because of this (at least for now). In most enterprise situations you simply can't ship apps that allow users to lose all their data by accidentally navigating away before saving. All frontend frameworks have client-side guards for this. 

I hope it gets solved soon - it's a pretty major issue. The solution would need to be purely client-side otherwise it'd really slow down `wire:navigate`.

**Reply by joshhanley** [reply]
@binaryfire and @codewithluck theres currently a PR #8093 in progress trying to work out the best way to approach this. Comments are welcome 🙂

**Reply by binaryfire** [reply]
@joshhanley Nice! That approach looks great.

**Reply by EPuigdemont** [reply]
You can intercept the navigate redirects if you're doing them from the Livewire component itself, but no dice when doing so from regular html.
Hopefully we get this navigating cancellable event, as I have found no way to interrupt the wire:navigating event.

**Reply by binaryfire** [reply]
@EPuigdemont Have a look at the PR Josh linked to.

**Reply by joshhanley** [reply, marked as answer]
PR #8093 has been merged with a `livewire:navigate` event that can be cancelled.
